ls1intum/ArTEMiS
# Feature/programming exercise/squash template commits + hard reset before structure oracle tests

### Checklist

### Motivation and Context



There are 2 workflow issues addressed in this PR:

1. When creating the template repository, the instructor/ta usually makes multiple commits onto it.
These commits are visible for the students when cloning their assignment repository (as it is a fork of the template repo). The instructors/tas therefore usually squash locally and then force push to the template repository.

2. This is a result of 1: When trying to update the oracle structure in programming-exercise > view, the endpoints fails as it can not pull from the remote repo as the remote branch history has changed.

### Description


1. I have introduced a button in programming-exercise > view that lets the instructor (and only the instructor) squash all template repo commits into the first commit.
This is done by soft resetting all commits but the first and then doing a commit --amend (unfortunately it is not possible in JGit to do a rebase --root, which would be the more straightforward solution).

2. Before pulling the template/solution/test repository for the generateOracleStructure endpoint, a hard reset to origin/master is executed.

Both actions should be used with caution:
:warning: They will reset all file changes that are unstaged/staged but not committed, as a hard reset to the remote is executed. This only concerns saved, but uncommitted changes from the instructor-view. :warning:

### Steps for Testing


For 1:

1. Log in to ArTEMiS
2. Open programming-exercise > view
3. Push commits from a local clone into the remote template repository
4. Click 'Squash template commits'
=> Check in Bitbucket, all commits should now be squashed into the first (all file changes are still there, but there is only the first commit)

For 2:

1. Log in to ArTEMiS
2. Open programming-exercise > view
3. Force Push commits from a local clone into the remote template/solution and test repository
4. Click 'Update oracle test structure'
=> Check in Bitbucket, should have successfully updated the oracle test structure

### Screenshots



![image](https://user-images.githubusercontent.com/28230611/59369255-4e254700-8d40-11e9-908e-16d09cc2400d.png)
